### PR TITLE
Fix issues with caching.

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/22176

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-query.php [hunk 1]
```diff
@@ -3454,7 +3454,7 @@ public function get_posts() {
 			/** @var WP_Post[] */
 			$this->posts = array_map( 'get_post', $this->posts );
 
-			if ( $q['cache_results'] ) {
+			if ( $q['cache_results'] && ! wp_using_ext_object_cache() ) {
 				update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
 			}
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-query.php]: Does this work as an alternative, it will only cache uncached post IDs.

`$post_ids = wp_list_pluck( $this->posts, 'ID' );` is run above but it would be good to confirm it's defined at this point under all possible branches.

```suggestion
			if ( $q['cache_results'] ) {
				_prime_post_caches( $post_ids, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
			}
```

File: src/wp-includes/class-wp-query.php [hunk 2]
```diff
@@ -3099,8 +3099,8 @@ public function get_posts() {
 				$last_changed .= wp_cache_get_last_changed( 'terms' );
 			}
 
-			$cache_key = "wp_query:$key:$last_changed";
-
+			$cache_key      = "wp_query:$key:$last_changed";
+			$cached_results = false;
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-query.php]:
```suggestion
			$cache_found = false;
```

Rather than presuming based on the return value, using the `$found` parameter would be better. This allows for falsey values to be returned without causing incorrect results. I think that can currently happen if the cached results are an empty array.

If it isn't possible now, it's a good defensive coding technique none-the-less.

:1234: The variable name will need to change throughout.

File: src/wp-includes/class-wp-query.php [hunk 3]
```diff
@@ -3099,8 +3099,8 @@ public function get_posts() {
...
 			if ( null === $this->posts ) {
 				$cached_results = wp_cache_get( $cache_key, 'posts' );
```
Comment (peterwilsoncc) [inline src/wp-includes/class-wp-query.php]:
```suggestion
				$cached_results = wp_cache_get( $cache_key, 'posts', false, $cache_found );
```

[`$cache_found` is passed by reference](http://developer.wordpress.org/reference/functions/wp_cache_get/) according to the docs.

**Review by peterwilsoncc — COMMENTED** [review]
I've added a few notes inline.

Something I've noticed with https://github.com/Automattic/advanced-post-cache is that it will cache random results too. I've been asking myself whether `$id_query_is_cacheable` should be filtered to allow developers to force caching of random queries anyway? Thoughts?

**Comment by peterwilsoncc** [comment]
Memcached failures look to be true failures https://github.com/WordPress/wordpress-develop/actions/runs/3139034236/jobs/5099033209#step:19:313

File: src/wp-includes/class-wp-query.php (hunk 3 above)
Comment (mukeshpanchal27) [reply src/wp-includes/class-wp-query.php]: +1

File: src/wp-includes/class-wp-query.php (hunk 2 above)
Comment (mukeshpanchal27) [reply src/wp-includes/class-wp-query.php]: Nitpick: I would have set this to '' (an empty string) for the default value, so the type of the variable doesn't change.

```suggestion
			$cache_found = '';
```

**Comment by mukeshpanchal27** [comment]
Left one nitpick suggestion https://github.com/WordPress/wordpress-develop/pull/3348#discussion_r981996311 Does it make sense @peterwilsoncc @spacedmonkey ?

File: src/wp-includes/class-wp-query.php (hunk 1 above)
Comment (mukeshpanchal27) [reply src/wp-includes/class-wp-query.php]: @peterwilsoncc I tested this changes but it break unit tests https://github.com/WordPress/wordpress-develop/pull/3353

Comment (spacedmonkey) [reply src/wp-includes/class-wp-query.php]: To honest, I am not even sure of the point of this line of code anyway. It is really confusing. Priming the caches conditionally seems weird when you already have `update_post_term_cache` and `update_post_meta_cache` for this. 

I would be happy removing `if ( $q['cache_results'] ) {` check all together. Thoughts @mukeshpanchal27 @peterwilsoncc ?

File: src/wp-includes/class-wp-query.php (hunk 2 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-query.php]: You know, I have never used the found param on this function. I like it. I have refactored the PR to use it.

File: src/wp-includes/class-wp-query.php (hunk 3 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-query.php]: +1. Pushed fix.

**Review by peterwilsoncc — APPROVED** [review]
These changes look good to me.

Do you think the following needs to added to the test suite? I don't think an empty result set is tested to ensure it's cached at present.

```php

/**
 * @ticket 22176
 * @dataProvider data_query_cache_with_empty_result_set
 */
public function test_query_cache_with_empty_result_set( $fields_q1, $fields_q2 ) {
	_delete_all_posts();

	$args_q1 = array(
		'fields'    => $fields_q1,
	);

	$query_1  = new WP_Query();
	$posts_q1 = $query_1->query( $args_q1 );
	$this->assertEmpty( $posts_q1, 'First query does not return an empty result set.' );

	$args_q2 = array(
		'fields'    => $fields_q2,
	);

	$num_queries = get_num_queries();
	$query_2     = new WP_Query();
	$posts_q2    = $query_2->query( $args_q2 );
	$this->assertEmpty( $posts_q2, 'Second query does not return an empty result set.' );
	$this->assertSame( $num_queries, get_num_queries(), 'Second query is not cached.' );
}

public function data_query_cache_with_empty_result_set() {
	return array(
		array( '', '' ),
		array( '', 'ids' ),
		array( '', 'id=>parent' ),

		array( 'ids', '' ),
		array( 'ids', 'ids' ),
		array( 'ids', 'id=>parent' ),

		array( 'id=>parent', '' ),
		array( 'id=>parent', 'ids' ),
		array( 'id=>parent', 'id=>parent' ),
	);
}
```

**Review by mukeshpanchal27 — APPROVED** [review]
LGTM. Peter's suggested unit tests also look good to include.

**Comment by spacedmonkey** [comment]
Committed

File: src/wp-includes/class-wp-query.php (hunk 1 above)
Comment (ocean90) [reply src/wp-includes/class-wp-query.php]: I think the switch to `_prime_post_caches()` wasn't a good idea, see https://core.trac.wordpress.org/ticket/57163.
